### PR TITLE
feat: cli subcommand for token metadata address extension

### LINKED ISSUE #4551: token-cli: Support metadata pointer extension
#### Problem

With https://github.com/solana-labs/solana-program-library/pull/4549, token-2022 can also store a pointer to a metadata account, but this isn't exposed in the CLI except for setting authorities.

#### Solution

Add the functionality

### PR DESCRIPTION
**Pull request by jim4067**
Resolves #4551

image

### CODE REVIEW

File: token/cli/src/main.rs [hunk 1]
```diff
@@ -2613,6 +2621,15 @@ fn app<'a, 'b>(
                             Rate authority defaults to the mint authority."
                         ),
                 )
+                .arg(
+                    Arg::with_name("metadata-address")
```
Comment (joncinque) [inline token/cli/src/main.rs:2625]: Otherwise it won't ever get picked up!
```suggestion
                    Arg::with_name("metadata_address")
```

File: token/cli/src/main.rs [hunk 2]
```diff
@@ -3768,8 +3785,10 @@ async fn process_command<'a>(
             let decimals = value_t_or_exit!(arg_matches, "decimals", u8);
             let mint_authority =
                 config.pubkey_or_default(arg_matches, "mint_authority", &mut wallet_manager)?;
-            let memo = value_t!(arg_matches, "memo", String).ok();
+            let memo: Option<String> = value_t!(arg_matches, "memo", String).ok();
```
Comment (joncinque) [inline token/cli/src/main.rs:3788]: nit: Can you revert this change?

File: token/cli/src/main.rs [hunk 3]
```diff
@@ -3768,8 +3785,10 @@ async fn process_command<'a>(
...
             let rate_bps = value_t!(arg_matches, "interest_rate", i16).ok();
+            let metadata_address: Option<Pubkey> =
+                value_t!(arg_matches, "metadata_address", Pubkey).ok();
```
Comment (joncinque) [inline token/cli/src/main.rs:3791]: nit for consistency
```suggestion
            let metadata_address = value_t!(arg_matches, "metadata_address", Pubkey).ok();
```

**Review by joncinque — COMMENTED** [review]
Can you also add a test for this? You just need to create the token with a metadata address and then check that the mint has that extension and metadata address.

As it stands, this doesn't work unfortunately, but with a test, we'll be sure!

File: token/cli/src/main.rs (hunk 2 above)
Comment (jim4067) [reply token/cli/src/main.rs:3788]: a push with the changes removed should do the trick, I think

**Comment by joncinque** [comment]
Let's keep all the conversation here rather than in two places. To build the programs, you need to go into `token/program-2022` and run `cargo build-sbf`, and then go into `token/program` and run `cargo build-sbf`, and then go into `associated-token-account/program` and run `cargo build-sbf`, then you can run `cargo test` in the CLI directory

File: token/cli/src/main.rs [hunk 4]
```diff
@@ -7425,4 +7453,64 @@ mod tests {
...
+            None,
+            Some(metadata_address),
+            None,
+            None,
+            None,
+            None,
+            bulk_signers,
+        )
+        .await
+        .unwrap();
```
Comment (joncinque) [inline token/cli/src/main.rs:7489]: Rather than using `command_create_token`, which doesn't exercise the command line parsing, can you execute using `process_test_command`, ie:

```
process_test_command(
    &config, &payer,
    &["spl-token", CommandName::CreateToken.into(), "--metadata-address", metadata_address.to_string()]
).await;
```

You're doing this later in the test, so can you remove this and move that call up here? You'll just need to extract the mint info, just like in the `create_token_default` test

Comment (jim4067) [reply token/cli/src/main.rs:7489]: Is it okay to omit the `--program_id` arguments?

Comment (joncinque) [reply token/cli/src/main.rs:7489]: Ah, no, definitely not

**Comment by jim4067** [comment]
having trouble testing. 
Getting this error when trying to build `program-2022` - `error: package `ark-serialize v0.4.2` cannot be built because it requires rustc 1.63 or newer, while the currently active rustc version is 1.62.0-dev`


I'm thinking it has something to do with the MSRV for these crates but I can't find them in the `cargo.toml`

**Comment by joncinque** [comment]
The rust toolchain is defined at the `rust-toolchain.toml` file https://github.com/solana-labs/solana-program-library/blob/master/rust-toolchain.toml

Otherwise, it looks like you're using an older version of the SBF platform tools. Be sure to run `solana-install init 1.16.1` to match up with the SPL repo

File: token/cli/src/main.rs [hunk 5]
```diff
@@ -7507,10 +7474,22 @@ mod tests {
...
+        .await;
+
+        let value: serde_json::Value = serde_json::from_str(&result.unwrap()).unwrap();
+        let mint = Pubkey::from_str(value["commandOutput"]["address"].as_str().unwrap()).unwrap();
+        let account = config.rpc_client.get_account(&mint).await.unwrap();
+        let mint_state = StateWithExtensionsOwned::<Mint>::unpack(account.data).unwrap();
+
+        let extension = mint_state.get_extension::<MetadataPointer>().unwrap();
+
+        assert!(mint_state.get_extension::<MetadataPointer>().is_ok());
```
Comment (joncinque) [inline token/cli/src/main.rs:7489]: nit: you can remove this line since you already unwrapped on the same call one line before

**Review by joncinque — COMMENTED** [review]
We're looking very close!

**Comment by jim4067** [comment]
let me know when it's ready, so that I can squash the commits.

File: token/cli/src/main.rs (hunk 5 above)
Comment (joncinque) [reply token/cli/src/main.rs:7488]: Sorry, I meant this exact line, not the whitespace that was there before

File: token/cli/src/main.rs [hunk 6]
```diff
@@ -7425,4 +7453,43 @@ mod tests {
...
         );
     }
+
+    #[tokio::test]
+    #[serial]
+    async fn metadata_pointer() {
+        let (test_validator, payer) = new_validator_for_test().await;
+        let config =
+            test_config_with_default_signer(&test_validator, &payer, &spl_token_2022::id());
+        let program_id = &spl_token_2022::id();
```
Comment (joncinque) [inline token/cli/src/main.rs:7463]: nit: you can reuse the same variable, no need to redeclare it elsewhere!
```suggestion
        let program_id = &spl_token_2022::id();
        let config =
            test_config_with_default_signer(&test_validator, &payer, &program_id);
```

File: token/cli/src/main.rs [hunk 7]
```diff
@@ -2613,6 +2621,15 @@ fn app<'a, 'b>(
                             Rate authority defaults to the mint authority."
                         ),
                 )
+                .arg(
+                    Arg::with_name("metadata_address")
+                        .long("metadata-address")
+                        .value_name("METADATA_ADDRESS")
```
Comment (joncinque) [inline token/cli/src/main.rs:2627]: nit: no need to show `--metadata-address METADATA_ADDRESS` in the help, so this can be simplified
```suggestion
                        .value_name("ADDRESS")
```

File: token/cli/src/main.rs [hunk 8]
```diff
@@ -2613,6 +2621,15 @@ fn app<'a, 'b>(
...
+                        .takes_value(true)
+                        .help(
+                            "Specify metadata address to use."
```
Comment (joncinque) [inline token/cli/src/main.rs:2630]: nit: let's make this a bit clearer
```suggestion
                            "Specify address that stores token metadata."
```

File: token/cli/src/main.rs [hunk 9]
```diff
@@ -7425,4 +7453,43 @@ mod tests {
...
     }
+
+    #[tokio::test]
+    #[serial]
+    async fn metadata_pointer() {
+        let (test_validator, payer) = new_validator_for_test().await;
+        let config =
+            test_config_with_default_signer(&test_validator, &payer, &spl_token_2022::id());
+        let program_id = &spl_token_2022::id();
+        let token_path = NamedTempFile::new().unwrap();
```
Comment (joncinque) [inline token/cli/src/main.rs:7464]: Is this needed? Only tests that need the file to be written somewhere use it. Otherwise, the CLI will generate a new keypair, and you can pull it out from `value["commandOutput"]["address"]`, as you're already doing

File: token/cli/src/main.rs [hunk 10]
```diff
@@ -7425,4 +7453,43 @@ mod tests {
...
+            &config,
+            &payer,
+            &[
+                "spl-token",
+                CommandName::CreateToken.into(),
+                token_path.path().to_str().unwrap(),
+                "--program-id",
+                &program_id.to_string(),
+                "--metadata-address",
+                metadata_address.to_string().as_str(),
```
Comment (joncinque) [inline token/cli/src/main.rs:7477]: nit for consistency with `program_id`
```suggestion
                &metadata_address.to_string(),
```

**Review by joncinque — COMMENTED** [review]
Sorry, since my last suggestion wasn't totally applied, I found a few other nits to clean up. All the functionality is correct though!

**Comment by joncinque** [comment]
> let me know when it's ready, so that I can squash the commits.

Don't worry, we have a "Squash and merge" policy by default, so feel free to add as many commit as you need

**Comment by joncinque** [comment]
Be sure to run `cargo fmt --all` too before committing

File: token/cli/src/main.rs (hunk 10 above)
Comment (jim4067) [reply token/cli/src/main.rs:7477]: Here we expect a string slice, that's why the `metadata_address.to_string().as_str()`

**Review by joncinque — COMMENTED** [review]
Sorry, one last bit!

File: token/cli/src/main.rs [hunk 11]
```diff
@@ -7425,4 +7453,39 @@ mod tests {
                 .unwrap()
         );
     }
+
+    #[tokio::test]
+    #[serial]
+    async fn metadata_pointer() {
+        let (test_validator, payer) = new_validator_for_test().await;
+        let program_id = &spl_token_2022::id();
+        let config = test_config_with_default_signer(&test_validator, &payer, &program_id);
```
Comment (joncinque) [inline token/cli/src/main.rs:7462]: Clippy's complaining about this, you can just do
```suggestion
        let program_id = spl_token_2022::id();
        let config = test_config_with_default_signer(&test_validator, &payer, &program_id);
```

**Review by joncinque — APPROVED** [review]
Thanks so much for your contribution!
